lib/src/utils/matrix_file.dart

```diff
@@ -233,12 +234,18 @@ class MatrixImageFile extends MatrixFile {
   /// you would likely want to use [NativeImplementations] and
   /// [Client.nativeImplementations] instead
   static MatrixImageFileResizedResponse? calcMetadataImplementation(
-      Uint8List bytes) {
-    final image = decodeImage(bytes);
+      MatrixImageFileCalcMetadataArguments arguments) {
+    Image? image;
+    try {
+      image = decodeNamedImage(arguments.fileName, arguments.bytes);
+    } catch (e) {
+      // An edge case where the image file is sent with a wrong extension.
+      image = decodeImage(arguments.bytes);
+    }
```

Comment:
Have you tested this against files with the wrong extension? And do you have any benchmarks to show the performance benefit?

Reply:
Actually, after checking the implementation, this does seem to only read the first 4-ish bytes of each file when trying to select the decoder, which doesn't seem to be much slower than comparing the 4-ish character file extensions. Specifically also that performance warning reads like this:

```
/// **WARNING** Since this will check the image data against all known decoders,
/// it is much slower than using an explicit decoder.
```

So it tells you to pass an explicit decoder, if you know the file format ahead of time. It does not say that decodeNamedImage would be way faster.

Reply:
Yes, looks like this doesn't improve the performance so much. But looks like it does perform a little bit better, a little worse sometimes too :( -

Attempt 1
```
-----------------------------------------------------------------------
| decodeImage               | decodeNamedImage ?? decodeImage       
-----------------------------------------------------------------------
0:00:00.612670              | 0:00:00.610485 
0:00:00.838736              | 0:00:00.825123 
0:00:01.281128              | 0:00:01.286197 
0:00:01.348387              | 0:00:01.305697 
0:00:01.303319              | 0:00:01.328091 
0:00:00.609605              | 0:00:00.720614 
0:00:00.980312              | 0:00:01.061956 
0:00:01.387428              | 0:00:01.379998 
0:00:01.304694              | 0:00:01.313241 
0:00:01.144865              | 0:00:01.151162 
0:00:00.599144              | 0:00:00.625925 
0:00:00.868126              | 0:00:00.859511 
0:00:01.405924              | 0:00:01.392055 
0:00:01.311831              | 0:00:01.309458 
0:00:01.142277              | 0:00:01.142348 
0:00:00.587615              | 0:00:00.591835 
0:00:00.943614              | 0:00:00.858286 
0:00:01.395859              | 0:00:01.386894 
0:00:01.306928              | 0:00:01.311965 
0:00:01.159670              | 0:00:01.135599 
0:00:00.585895              | 0:00:00.582027 
0:00:01.019137              | 0:00:00.840622 
0:00:01.692277              | 0:00:01.429335 
0:00:01.303378              | 0:00:01.306888 
0:00:01.150505              | 0:00:01.144645 
0:00:00.592935              | 0:00:00.585403 
0:00:00.877589              | 0:00:00.890118 
0:00:01.391462              | 0:00:01.390174 
0:00:01.302328              | 0:00:01.314887 
0:00:01.153736              | 0:00:01.163653 
0:00:00.598910              | 0:00:00.594785 
0:00:00.870141              | 0:00:00.859396 
0:00:01.406210              | 0:00:01.389826 
0:00:01.309853              | 0:00:01.308013 
0:00:01.143563              | 0:00:01.137913 
0:00:00.606268              | 0:00:00.614362 
0:00:00.863765              | 0:00:00.929814 
0:00:01.768525              | 0:00:01.391816 
0:00:01.517358              | 0:00:01.314393 
0:00:01.141153              | 0:00:01.146730 
0:00:00.597237              | 0:00:00.585827 
0:00:00.863349              | 0:00:00.882129 
0:00:01.449131              | 0:00:01.387415 
0:00:01.303109              | 0:00:01.315909 
0:00:01.145529              | 0:00:01.136813 
0:00:00.578354              | 0:00:00.593698 
0:00:00.857400              | 0:00:00.851482 
0:00:01.387541              | 0:00:01.408726 
0:00:01.322078              | 0:00:01.317221 
0:00:01.161776              | 0:00:01.141493 
-----------------------------------------------------------------------
(29/50) times decodeNamedImage ?? decodeImage performed better
```

Attempt 2

```
-----------------------------------------------------------------------
| decodeImage               | decodeNamedImage ?? decodeImage       
-----------------------------------------------------------------------
0:00:00.632311              | 0:00:00.619077 
0:00:00.838749              | 0:00:00.808607 
0:00:01.293018              | 0:00:01.280403 
0:00:01.340220              | 0:00:01.311602 
0:00:01.182309              | 0:00:01.151103 
0:00:00.605361              | 0:00:00.591334 
0:00:00.896182              | 0:00:00.849899 
0:00:01.384815              | 0:00:01.370898 
0:00:01.303706              | 0:00:01.317590 
0:00:01.136142              | 0:00:01.144172 
0:00:00.572908              | 0:00:00.587720 
0:00:01.008201              | 0:00:01.057481 
0:00:01.384842              | 0:00:01.371472 
0:00:01.323096              | 0:00:01.335359 
0:00:01.158683              | 0:00:01.150877 
0:00:00.597109              | 0:00:00.608542 
0:00:00.967329              | 0:00:00.843949 
0:00:01.395706              | 0:00:01.371828 
0:00:01.310032              | 0:00:01.318538 
0:00:01.152430              | 0:00:01.143048 
0:00:00.571923              | 0:00:00.588044 
0:00:00.843148              | 0:00:00.852557 
0:00:01.386903              | 0:00:01.376499 
0:00:01.312540              | 0:00:01.324892 
0:00:01.147622              | 0:00:01.140033 
0:00:00.571529              | 0:00:00.600541 
0:00:01.098011              | 0:00:00.967342 
0:00:01.382154              | 0:00:01.368603 
0:00:01.298306              | 0:00:01.324912 
0:00:01.145135              | 0:00:01.149384 
0:00:00.600829              | 0:00:00.583516 
0:00:00.939102              | 0:00:00.888255 
0:00:01.378334              | 0:00:01.372755 
0:00:01.297905              | 0:00:01.293041 
0:00:01.140712              | 0:00:01.135706 
0:00:00.601973              | 0:00:00.603937 
0:00:00.910251              | 0:00:00.842211 
0:00:01.380103              | 0:00:01.361315 
0:00:01.299755              | 0:00:01.311302 
0:00:01.146147              | 0:00:01.143558 
0:00:00.586058              | 0:00:00.753099 
0:00:01.077423              | 0:00:00.844978 
0:00:01.565908              | 0:00:01.544572 
0:00:01.299474              | 0:00:01.323327 
0:00:01.150291              | 0:00:01.141532 
0:00:00.592465              | 0:00:00.589737 
0:00:00.854104              | 0:00:00.869504 
0:00:01.400851              | 0:00:01.385930 
0:00:01.325397              | 0:00:01.321316 
0:00:01.142043              | 0:00:01.158076 
-----------------------------------------------------------------------
(31/50) times decodeNamedImage ?? decodeImage performed better
```

Reply:
In case you wanted to run this. I just removed one image from the directory so I can upload it under 25MB.

[AlphanumericStimulatingLanguage.zip](https://github.com/famedly/matrix-dart-sdk/files/15296872/AlphanumericStimulatingLanguage.zip)

Reply:
Attempt 3

```
-----------------------------------------------------------------------
| decodeImage               | decodeNamedImage ?? decodeImage       
-----------------------------------------------------------------------
0:00:00.617083              | 0:00:00.585860 
0:00:00.834217              | 0:00:00.806740 
0:00:01.335568              | 0:00:01.359266 
0:00:01.177358              | 0:00:01.152674 
0:00:00.606037              | 0:00:00.585520 
0:00:00.847492              | 0:00:00.837612 
0:00:01.333227              | 0:00:01.303109 
0:00:01.148316              | 0:00:01.147245 
0:00:00.585660              | 0:00:00.575768 
0:00:00.879510              | 0:00:00.842727 
0:00:01.316270              | 0:00:01.314509 
0:00:01.148291              | 0:00:01.151987 
0:00:00.570452              | 0:00:00.579830 
0:00:00.855215              | 0:00:00.920261 
0:00:01.409951              | 0:00:01.334502 
0:00:01.143889              | 0:00:01.155418 
0:00:00.594920              | 0:00:00.586448 
0:00:00.874756              | 0:00:00.834236 
0:00:01.306674              | 0:00:01.319978 
0:00:01.151949              | 0:00:01.149007 
0:00:00.637402              | 0:00:00.578597 
0:00:00.856722              | 0:00:00.847611 
0:00:01.312554              | 0:00:01.324049 
0:00:01.148258              | 0:00:01.148096 
0:00:00.603520              | 0:00:00.582952 
0:00:00.855248              | 0:00:00.842612 
0:00:01.316368              | 0:00:01.318722 
0:00:01.156176              | 0:00:01.149882 
0:00:00.605124              | 0:00:00.607661 
0:00:00.873450              | 0:00:01.018019 
0:00:01.308119              | 0:00:01.305516 
0:00:01.148917              | 0:00:01.298632 
0:00:00.610257              | 0:00:00.587137 
0:00:00.874313              | 0:00:00.917965 
0:00:01.410471              | 0:00:01.335019 
0:00:01.147080              | 0:00:01.135510 
0:00:00.587952              | 0:00:00.642588 
0:00:00.860974              | 0:00:00.860354 
0:00:01.347320              | 0:00:01.326533 
0:00:01.156110              | 0:00:01.148404 
-----------------------------------------------------------------------
(27/40) times decodeNamedImage ?? decodeImage performed better
```

Reply:
That is just measurement inaccuracies, I would assume. If you do benchmarks it makes sense to run measurements multiple times and calculate an error bar for the measurement error. There are some libraries that can help with that.